R/build-search-docs.R

```diff
@@ -65,3 +68,138 @@ url_node <- function(url) {
     paste0("<url><loc>", url, "</loc></url>")
   )
 }
+
+#' Build search index
+#'
+#' Build a JSON file encompassing all HTML pages, for use by the search script.
+#'
+#' @section YAML config:
+#' You can exclude some paths from the search index.
+#' Below we exclude the changelog from the search index.
+#'
+#' ```yaml
+#' search:
+#'   exclude: ['news/index.html']
+#' ```
+#' @section Local testing:
+#' Search won't work if you simply use pkgdown preview of the static files.
+#' You can use `servr::httw("docs")` instead.
+#'
+#'
+#' @inheritParams build_articles
+#' @export
+#'
+build_search <- function(pkg = ".",
+                         override = list()) {
+  pkg <- section_init(pkg, depth = 1L, override = override)
+  rule("Building search index")
+
+  paths <- fs::path_rel(
+      fs::dir_ls(pkg$dst_path, glob = "*.html", recurse = TRUE),
+      pkg$dst_path
+  )
+  paths <- paths[!paths %in% c("404.html", "articles/index.html", "reference/index.html")]
+
+  # user-defined exclusions
+  paths <- paths[!paths %in% pkg$meta$search$exclude]
+
+  index <- lapply(paths, file_search_index, pkg = pkg)
+  index <- unlist(index, recursive = FALSE)
+  index <- purrr::compact(index)
+  jsonlite::write_json(
+    index,
+    file.path(pkg$dst_path, "search.json"),
+    auto_unbox = TRUE
+  )
+}
+
+file_search_index <- function(path, pkg) {
+  html <- xml2::read_html(file.path(pkg$dst_path, path))
+  # Get page title
+  title <- xml2::xml_find_first(html, ".//meta[@property='og:title']") %>%
+    xml2::xml_attr("content")
+
+  # Get contents minus logo
+  node <- xml2::xml_find_all(html, ".//div[contains(@class, 'contents')]")
+  xml2::xml_remove(xml2::xml_find_first(node, ".//img[contains(@class, 'pkg-logo')]"))
+  sections <- xml2::xml_find_all(node, ".//*[contains(@class, 'section')]")
+
+  # Function for extracting, if it exists, the nearest h2 (needed for the changelog)
```

Comment:
Don't we always want it to show on the search results?

Reply:
Do you mean that for any results, we should show all the breadcrumbs leading to its heading?

Reply:
I'll experiment in this direction.

Reply:
I think that would be ideal, if it's not too much work.